gcc/rust/parse/rust-parse.cc

```diff
@@ -150,50 +151,51 @@ peculiar_fragment_match_compatible_fragment (
   return is_valid;
 }
 
+// The following function computes the follow-set for the :vis fragment.
+static std::vector<TokenId>
+vis_fragment_follow_set ()
+{
+  std::vector<TokenId> follow_set = {
+    COMMA,	   IDENTIFIER /* FIXME: Other than `priv` */,
+    LEFT_PAREN,	   LEFT_SQUARE,
+    EXCLAM,	   ASTERISK,
+    AMP,	   LOGICAL_AND,
+    QUESTION_MARK, LIFETIME,
+    LEFT_ANGLE,	   LEFT_SHIFT,
+    UNDERSCORE,
+    // FIXME: Add $crate as valid
+  };
+  // Add all keyword identifiers.
+  // RS_TOKEN_LIST is defined in rust-token.h.
+#define RS_TOKEN(name, _)
+#define RS_TOKEN_KEYWORD(x, _)                                                 \
+  if (x != PRIV)                                                               \
+    {                                                                          \
+      follow_set.push_back (x);                                                \
+    }
+  RS_TOKEN_LIST
+#undef RS_TOKEN_KEYWORD
+#undef RS_TOKEN
+  return follow_set;
+}
+
 static bool
 peculiar_fragment_match_compatible (const AST::MacroMatchFragment &last_match,
 				    const AST::MacroMatch &match)
 {
   static std::unordered_map<AST::MacroFragSpec::Kind, std::vector<TokenId>>
-    follow_set
-    = {{AST::MacroFragSpec::EXPR, {MATCH_ARROW, COMMA, SEMICOLON}},
-       {AST::MacroFragSpec::STMT, {MATCH_ARROW, COMMA, SEMICOLON}},
-       {AST::MacroFragSpec::PAT, {MATCH_ARROW, COMMA, EQUAL, PIPE, IF, IN}},
-       {AST::MacroFragSpec::PATH,
-	{MATCH_ARROW, COMMA, EQUAL, PIPE, SEMICOLON, COLON, RIGHT_ANGLE,
-	 RIGHT_SHIFT, LEFT_SQUARE, LEFT_CURLY, AS, WHERE}},
-       {AST::MacroFragSpec::TY,
-	{MATCH_ARROW, COMMA, EQUAL, PIPE, SEMICOLON, COLON, RIGHT_ANGLE,
-	 RIGHT_SHIFT, LEFT_SQUARE, LEFT_CURLY, AS, WHERE}},
-       {AST::MacroFragSpec::VIS,
-	{
-	  COMMA,
-	  IDENTIFIER /* FIXME: Other than `priv` */,
-	  LEFT_PAREN,
-	  LEFT_SQUARE,
-	  EXCLAM,
-	  ASTERISK,
-	  AMP,
-	  LOGICAL_AND,
-	  QUESTION_MARK,
-	  LIFETIME,
-	  LEFT_ANGLE,
-	  LEFT_SHIFT,
-	  SUPER,
-	  SELF,
-	  SELF_ALIAS,
-	  EXTERN_TOK,
-	  CRATE,
-	  UNDERSCORE,
-	  FOR,
-	  IMPL,
-	  FN_TOK,
-	  UNSAFE,
-	  TYPEOF,
-	  DYN
-	  // FIXME: Add Non kw identifiers
-	  // FIXME: Add $crate as valid
-	}}};
+    follow_set = {
+      {AST::MacroFragSpec::EXPR, {MATCH_ARROW, COMMA, SEMICOLON}},
+      {AST::MacroFragSpec::STMT, {MATCH_ARROW, COMMA, SEMICOLON}},
+      {AST::MacroFragSpec::PAT, {MATCH_ARROW, COMMA, EQUAL, PIPE, IF, IN}},
+      {AST::MacroFragSpec::PATH,
+       {MATCH_ARROW, COMMA, EQUAL, PIPE, SEMICOLON, COLON, RIGHT_ANGLE,
+	RIGHT_SHIFT, LEFT_SQUARE, LEFT_CURLY, AS, WHERE}},
+      {AST::MacroFragSpec::TY,
+       {MATCH_ARROW, COMMA, EQUAL, PIPE, SEMICOLON, COLON, RIGHT_ANGLE,
+	RIGHT_SHIFT, LEFT_SQUARE, LEFT_CURLY, AS, WHERE}},
+      {AST::MacroFragSpec::VIS, vis_fragment_follow_set ()},
```

Comment:
I can see you are trying to avoid having duplication I am starting to wonder for this patch would it not just be simpler to just add the keywords here?